File: JavaScript/KillProcess/KillProcess.js
/**
 * (465ms)
 * @param {number[]} pid
 * @param {number[]} ppid
 * @param {number} kill
 * @return {number[]}
 */
var killProcess = function(pid, ppid, kill) {
    var kids = new Map();
    for(var i = 0; i < ppid.length; i++) {
    	if(kids.has(ppid[i])) {
    		kids.get(ppid[i]).push(pid[i]);
    	} else {
    		kids.set(ppid[i], [pid[i]]);
    	}
    }
    var queue = [kill], ans = [];
    while(queue.length > 0) {
    	var cur = queue.shift();
    	ans.push(cur);
    	if(kids.has(cur)) {
    		queue = queue.concat(kids.get(cur));
    	}
    }
    return ans;
};

File: Python/KillProcess/KillProcess.py
class Solution(object):
	# (579ms)
    def killProcess(self, pid, ppid, kill):
        """
        :type pid: List[int]
        :type ppid: List[int]
        :type kill: int
        :rtype: List[int]
        """
        kids = {}
        for index, num in enumerate(ppid):
        	if num in kids:
        		kids[num].append(pid[index])
        	else:
        		kids[num] = [pid[index]]

        queue, ans = [kill], []
        while queue:
        	cur = queue.pop()
        	ans.append(cur)
        	if cur in kids:
        		queue += kids[cur]
        return ans

pid, ppid = [1, 3, 5, 10], [3, 0, 3, 5]
print Solution().killProcess(pid, ppid, 5)
